### DISCUSSION TITLE
How to load dynamiclib automatically without invoke dlopen+dlsym?

### DISCUSSION THREAD

**Discussion by oatgnauh** [emscripten-core/emscripten#22398] [Q&A] [answered]
emsdk 3.1.54
multi thread browser environment
Now I download sideModule into FS and passing fs path to `dlopen`, then use `dlsym` to load apis，I wonder if there is a better way to do it automatically.

I add `        " \
        -sMAIN_MODULE=2 \
        -sAUTOLOAD_DYLIBS=1 \
        ${ffmpeg_DIR}/lib/libffmpeg${WASM_SUBFFIX_NAME}.wasm \
        "`
to main module link flag, according to [official doc](https://emscripten.org/docs/compiling/Dynamic-Linking.html#load-time-dynamic-linking), it should be load-time dynamic link

**Comment by oatgnauh** [comment]
I learn something from
![image](https://github.com/user-attachments/assets/3bce9993-1949-41a1-a3ff-94c5c13d1b4b)

`AUTOLOAD_DYLIBS` just help me download simd module, can not do dlopen&dlsym

also when I review this js code:
```javascript
var loadDylibs = () => {
 if (!dynamicLibraries.length) {
  reportUndefinedSymbols();
  return;
 }
 addRunDependency("loadDylibs");
 dynamicLibraries.reduce((chain, lib) => chain.then(() => loadDynamicLibrary(lib, {
  loadAsync: true,
  global: true,
  nodelete: true,
  allowUndefined: true
 })), Promise.resolve()).then(() => {
  reportUndefinedSymbols();
  removeRunDependency("loadDylibs");
 });
};
```

seems if you want to do load-time dynamick link, you have to set the array `dynamicLibraries`, which is define at `var dynamicLibraries = Module["dynamicLibraries"] || [];`

but still have some problem. I set 
`Module["dynamicLibraries"] = ["libffmpeg_simd.wasm"]`
and i got a crash



still confusing..

**Comment by oatgnauh** [comment]
however, I found that even if i am using load-time link,  it would invoke `loadDynamicLibrary` automatically by `createWasm -> loadDylibs ->loadDynamicLibrary`, but C function symbols are not being export.

here is dlsym function in js
```javascript
var __dlsym_js = (handle, symbol, symbolIndex) => {
 symbol = UTF8ToString(symbol);
 var result;
 var newSymIndex;
 var lib = LDSO.loadedLibsByHandle[handle];
 if (!lib.exports.hasOwnProperty(symbol) || lib.exports[symbol].stub) {
  dlSetError(`Tried to lookup unknown symbol "${symbol}" in dynamic lib: ${lib.name}`);
  return 0;
 }
 newSymIndex = Object.keys(lib.exports).indexOf(symbol);
 result = lib.exports[symbol];
 if (typeof result == "function") {
  var addr = getFunctionAddress(result);
  if (addr) {
   result = addr;
  } else {
   result = addFunction(result, result.sig);
   GROWABLE_HEAP_U32()[((symbolIndex) >> 2)] = newSymIndex;
  }
 }
 return result;
};
```
most importance call is `getFunctionAddress` and `addFunction`, none of two have been invoke in `loadDynamicLibrary`. 
So I guess this "load-time linking" simply do stuff as `dlopen` ?

**Comment by oatgnauh** [comment, marked as answer]
thanks myself, I figure out `Module["dynamicLibraries"] = ["libffmpeg_simd.wasm"]` should be add in pre.js, everything work.
remember add `-sAUTOLOAD_DYLIBS=1 `

**Reply by sbc100** [reply]
`AUTOLOAD_DYLIBS` is enabled by default so should need to pass `-sAUTOLOAD_DYLIBS=1`

When you pass `libffmpeg_simd.wasm` on the command line when linking your main module then `libffmpeg_simd.wasm` should automatically be added `dynamicLibraries`.. you should not need to add it manually anywhere.

**Reply by oatgnauh** [reply]
seems something wrong, `ibffmpeg_simd.wasm` bas been pass to command line but `Module["dynamicLibraries"] = ["libffmpeg_simd.wasm"]` not add to js file

here is default output
```javascript
var dynamicLibraries = Module["dynamicLibraries"] || [];
```

**Reply by oatgnauh** [reply]
I search the whole project
![image](https://github.com/user-attachments/assets/9f0a835f-5ecc-4e6e-a1fb-d0850098f3b3)

only notice that in `upstream/emscripten/test/test_core.py` 
![image](https://github.com/user-attachments/assets/e2ae6536-162f-4079-b0c0-2a2cc20732c8)
would create an pre.js including `Module['dynamicLibraries'] = ['liblib.so']`.
IIUC, this is just a test file

**Reply by sbc100** [reply]
If you look at your main module it should contain `dylink` header which includes the libraries it depends on:

```
$ wasm-objdump -x a.out.wasm
a.out.wasm:     file format wasm 0x1

Section Details:

Custom:
 - name: "dylink.0"
 - mem_size     : 4688
 - mem_p2align  : 4
 - table_size   : 5
 - table_p2align: 0
 - needed_dynlibs[1]:
  - libside.wasm                    <--- here you should see `libffmpeg_simd.wasm`
```

The emscripten startup code will add these libraries to `Module['dynamicLibraries'] ` on startup: https://github.com/emscripten-core/emscripten/blob/7f8a05dd4e37cbd7ffde6d624f91fd545f7b52e3/src/preamble.js#L957-L961

Can you check this code fragment is in your generated code?

**Reply by oatgnauh** [reply]
True, you are right
```txt
Custom:
 - name: "dylink.0"
 - mem_size     : 22830136
 - mem_p2align  : 4
 - table_size   : 77119
 - table_p2align: 0
 - needed_dynlibs[1]:
  - libffmpeg_simd.wasm
```

![image](https://github.com/user-attachments/assets/a19e19b7-a8e2-4bd4-8057-240da4b5fbb6)
Module["dynamicLibraries"] dont need to add manually
